### ISSUE TITLE
Reproducing Results (Table 1)

### ISSUE THREAD

**Issue by nil123532** [ptrckhmmr/learning-to-defer-with-limited-expert-predictions#3]
Dear Authors,

I have a question regarding your evaluation methodology. In the paper, you mention that “for each _l,_ we draw instances randomly from the training data while ensuring balanced class proportions.” However, from the code, it appears that the evaluation is performed using [dlval](https://github.com/ptrckhmmr/learning-to-defer-with-limited-expert-predictions/blob/07525da6f57a3bb2c4e8aa6f32577e806bd983a5/Embedding-Semi-Supervised/Train_embedding_fm.py#L405), which is built from data that is disjoint from the training set. Could you please clarify how this is intended to work and how one should reproduce Table 1 in your paper?

**Comment by lukasthede** [comment]
Dear nil123532,

To analyze the tradeoff between labeling costs and performance, we vary the number of available labels in our experiments. To ensure an equal number of labeled samples per class, we divide the total number of labels by the number of classes and randomly select the required number of samples for each class. You can find the corresponding implementation in lines 103–118 of [Semi-Supervised/datasets/cifar.py](https://github.com/ptrckhmmr/learning-to-defer-with-limited-expert-predictions/blob/main/Semi-Supervised/datasets/cifar.py).

Table 1 in our paper provides an overview of the quality of the generated artificial expert labels. Specifically, we compute the F-0.5 score between the artificial expert labels and the ground-truth expert labels on the training set. This metric helps assess how well the artificial expert labels approximate the true expert labels.

Let me know if you have any further questions!

Best,
Lukas

**Comment by nil123532** [comment]
Thank you for your prompt response both here and via email.

From my understanding, once training is complete, you generate the expert labels from the binary output and then evaluate using the F0.5 score. Could you please let me know if you have code available for that evaluation? Also, could you clarify which dataset you use for F0.5 score evaluation: the test set or a combination of the labeled and unlabeled training sets?


Best,
Nilesh Ramgolam

**Comment by lukasthede** [comment]
Hi Nilesh,

That’s correct. We evaluate the artificial expert labels by computing the F0.5 score between the artificial and true expert labels on the test set of CIFAR-100. While we do not have dedicated code for this in our repository, computing the F0.5 score is straightforward—for example, you can use the fbeta_score function from the sklearn library.

Let me know if you need further clarification!

Best,
Lukas

**Comment by nil123532** [comment]
Thank you, 
I shall continue my quest of reproducing results!

**Comment by nil123532** [comment]
Hi,

I have a few additional clarifications. 
It appears that the artificially generated label is binary ouput; indicating whether the expert is correct or not. 
Do we then use that artificially generated binary labels and dlval (which also has binary expert labels) to get the F0.5 score?

Thanks.

**Comment by nil123532** [comment]
Hello,

I hope you’re doing well. I noticed that in the embedding model’s learning rate graph, the LR quickly decays to 8e‑4, which might be due to the scheduler step being called after every mini‐batch (on line [99](https://github.com/ptrckhmmr/learning-to-defer-with-limited-expert-predictions/blob/07525da6f57a3bb2c4e8aa6f32577e806bd983a5/Embedding-Semi-Supervised/feature_extractor/emb_model_lib.py#L99) ). As a result, the schedule reaches the milestone of 160 steps almost immediately. It might be more appropriate to call the scheduler’s step() method once per epoch, so the learning rate decays at the intended intervals.


![Image](https://github.com/user-attachments/assets/e509916d-4f77-44f8-9cbc-1cbc35350383)

Because of this rapid LR decay, the model only reached around 63% accuracy on CIFAR, which is relatively low for an EfficientNet‐based approach. I made a small change to increment the scheduler’s epoch counter only after each epoch, instead of after every batch.

Separately, regarding the training of the “expertise” model (a linear model), I noticed that loss_x decreases while loss_u increases and then remains at that level for the rest of the 50 epochs. I would appreciate any guidance on how best to address this issue.

![Image](https://github.com/user-attachments/assets/775dc274-429b-4d84-8ccb-32653a3fbe06)

At the moment, I suspect it's because the embedding model might not have been trained properly, and that could explain why I’m seeing an F0.5 score of 74% (for n_labelled = 120) instead of 84% for Embedding‐FixMatch. Any advice or suggestions would be greatly appreciated. 
I have tried with the idea of stepping the scheduler after an epoch while training the Embedding model, then val accuracy of the Embedding Model increases from 63% to 80%. While, the embedding model's accuracy increases, I still get 76% for the F0.5 score when evaluating the expertise model and not the desired 84% when n_labelled=120.

Thank you, and I look forward to your response.

**Comment by nil123532** [comment]
This function evaluates the F0.5 score for a binary classification task using `sklearn`'s `fbeta_score`. The dataloader's labels are binary - whether the expert is correct or not

```python
import torch
from sklearn.metrics import fbeta_score

def evaluate_f0_5_sklearn(model, ema_model, emb_model, dataloader, beta=0.5):
    """
    Evaluate the F0.5 score on a binary classification (expert correct vs. incorrect)
    using sklearn's fbeta_score.

    :param model: Trained model for predictions.
    :param ema_model: Optional EMA model; if provided, it is used for predictions.
    :param emb_model: Embedding model that provides get_embedding().
    :param dataloader: DataLoader yielding (images, labels, im_ids).
                       Labels are binary: 1 if expert is correct, 0 otherwise.
    :param beta: Beta value for fbeta_score (default 0.5).
    :return: Computed F0.5 score (float).
    """
    model.eval()
    y_true, y_pred = [], []

    with torch.no_grad():
        for ims, lbs, _ in dataloader:
            ims, lbs = ims.cuda(), lbs.cuda()

            # Extract embeddings and make predictions
            embedding = emb_model.get_embedding(batch=ims)
            logits = ema_model(embedding) if ema_model else model(embedding)

            # Convert logits to predicted labels (0 or 1)
            output = torch.softmax(logits, dim=1)
            predicted_class = torch.argmax(output, dim=1)

            # Collect ground truth and predictions
            y_true.extend(lbs.cpu().tolist())
            y_pred.extend(predicted_class.cpu().tolist())

    return fbeta_score(y_true, y_pred, beta=beta)
```

**Comment by lukasthede** [comment]
Hi Nilesh,

To compute the F0.5 scores, which assess how closely the artificially generated expert labels align with the true expert labels, we first need to generate the artificial expert labels. Below is a step-by-step guide on how to achieve this.

### Step 1: Simulate the True Expert Labels
The first step is to generate the true expert labels for CIFAR by simulating a human expert with a specified level of “strength.” This can be done using the following command:
```
python generate_synthetic_experts.py --num_classes 20 --strength 60
```
which will save the true expert labels in `{artificial_ex_labels_path}/TrueEx_cifar100_{args.strength}_labels.json`.

### Step 2: Learn to Mimic the Expert
Once the true expert labels are generated, we train a model to mimic the expert in order to generate artificial expert labels. This process consists of two main steps:

#### 2.1 Train the Embedding Model

We first train an embedding model to produce meaningful representations of the input data, which will later serve as the foundation for approximating the true expert labels. The embedding model can be trained using:
```
python Train_emb_model.py --dataset cifar100 --model efficientnet_b1 --num_classes 20 --lr 0.1
```
Here, we use an *efficientnet_b1* model pretrained on ImageNet-1k from the timm library. You can evaluate the quality of your trained embedding model by performing a nearest neighbor classification in the learned embedding space.

#### 2.2 Train the Expert Model
Next, we train a model to mimic the true expert labels. We employ semi-supervised learning techniques (CoMatch and FixMatch) to leverage both labeled data (with true expert labels) and unlabeled data. You can train the expert model using either of the following commands:
```
python Train_embedding_cm.py --n-labeled 120 --ex_strength 60 --dataset CIFAR100
```
or 
```
python Train_embedding_fm.py --n-labeled 120 --ex_strength 60 --dataset CIFAR100
```
depending on whether you prefer to use CoMatch or FixMatch.

These scripts will first train the expert model using the true expert labels generated in Step 1. Once training is complete, the model will generate artificial expert labels for all samples in the training and test sets, which will be stored in the `artificial_expert_labels/` directory.


### Step 3: Assess the Quality of the Artificial Expert Labels
To evaluate the quality of the generated artificial expert labels, we compute the F0.5 scores by comparing them with the true expert labels on the test set. Both the true and generated artificial expert labels can be loaded from the `artificial_expert_labels/` directory.

I hope this provides clarity. Let me know if you have any further questions!

Best,
Lukas

**Comment by nil123532** [comment]
Thank you for your detailed response. 
I have some clarifications to do:

The file TrueEx_cifar100_60_labels.json was generated by running:
```
python generate_synthetic_experts.py --num_classes 20 --strength 60 
```
![Image](https://github.com/user-attachments/assets/41cdf70d-11fb-49f1-8ca7-a0d7a36623e7)
The file will contain labels from 0-19 classes 

And from this command:

``` 
python Train_embedding_fm.py --n-labeled 120 --ex_strength 60 --dataset CIFAR100
```
This following .json file is produced whose entries are binary indicators showing whether the expert is correct on each image:

![Image](https://github.com/user-attachments/assets/1945fc43-ee58-407b-8ba8-3e66a88cb838)


Because these two files represent data differently (one with multiclass labels 0–20, the other with binary correct/incorrect indicators), you have two options for computing an F0.5 score:

1. Binary Classification Approach:
- Use the binary outputs directly (expert correct vs. not correct) to calculate the F0.5 score. (Of course we will also need a .json where we have ground truth of expert being correct or not)

2. Multiclass Approach:
- Use the `preprocess_binary_predictions.py` script in Learning-to-Defer-Algs to convert the binary outputs into 0–19 class labels(see .json below). You can then apply a multiclass F0.5 score (e.g., using a weighted average) with scikit-learn’s [fbeta_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.fbeta_score.html).

![Image](https://github.com/user-attachments/assets/05d9b8ca-3fe1-4ff2-a017-4625f10ffb16)

Neither of these F0.5 Score gives me the desired 84.06% accuracy reported on the paper from this setting

Grateful if you could attend to my queries asap 

Thank you,
Nilesh

**Comment by nil123532** [comment]
![Image](https://github.com/user-attachments/assets/3b86d016-2876-4580-8a2b-02d2cd79f5f8)

I forgot to mention that the posterize_func in randaugment.py isn't working, so I've replaced it with torch.transform to perform random augmentation in cifar.py. Could this affect performance?

**Comment by lukasthede** [comment]
Hi Nilesh,

In our analysis, we compute the F0.5 scores between the binary true expert labels and the binary artificial expert labels.

Thank you for pointing out the issue in posterize_func within RandAugment. You can easily fix this error by ensuring that bits is clamped within a valid range:
```
def posterize_func(img, bits):
    bits = np.clip(bits, 1, 8)  # Ensure bits is in the range [1, 8]
    out = np.bitwise_and(img, np.uint8(255 << (8 - bits)))
    return out
```
We recommend using the augmentations as implemented in the provided codebase, as these match the original implementations from FixMatch and CoMatch.

**Troubleshooting Steps for Replicating Our Results**

If you’re experiencing issues reproducing our results, the following steps may help identify potential discrepancies:
* Check the embedding model performance: In our experiments on CIFAR-100, the final embedding model achieved a validation accuracy of 87.88%. Ensure your model reaches a similar level of performance.
* Verify hyperparameter settings: Double-check that the hyperparameters for training the expertise model are correctly set. You can find the exact values we used in Appendix Tables B1 and B2.
* Ensure proper seed initialization: Make sure the random seeds are correctly set both when simulating the CIFAR expert and when training the expert model to ensure reproducibility.

Let me know if you need further clarification!

Best,
Lukas

**Comment by nil123532** [comment]
Ok! Thanks I feel like the issue is with my embedding model then, because i am getting an max val accuracy of 80%

**Comment by lukasthede** [comment]
Hi Nilesh,

If the embedding model is not achieving the expected accuracy, you may want to check the following:
* Verify that pretrained weights are loaded correctly: We use an ImageNet-pretrained EfficientNet_B1 (loaded from *timm*), which we fine-tune on CIFAR-100 using the coarse classes (i.e., num_classes = 20).
* Ensure hyperparameters are correctly set: We train the embedding model using SGD with step-wise learning rate scheduling, a batch size of 64, and a total of 200 epochs.

Additionally, note that you can use alternative models for embeddings (e.g., ViT). The key requirement is that the embedding model generates meaningful representations of CIFAR-100 samples, which we then leverage in later stages to learn how to mimic the expert.

Let me know if you have any further questions!

Best,
Lukas

**Comment by nil123532** [comment]
Hi There,
Thank you for your response. 

I have managed to replicate results for FIXMATCH only. 

But am still struggling with the embedding model achieving only 80% accuracy on validation dataset after  all the specific hyper parameters are fixed

**Comment by nil123532** [comment]
Hello,

I was able to reproduce the results. However, there appears to be an issue with how the scheduler is stepped in [Train_emb_model.py](https://github.com/ptrckhmmr/learning-to-defer-with-limited-expert-predictions/blob/main/Embedding-Semi-Supervised/Train_emb_model.py): it should occur after every epoch. I’ve also found that using [ReduceLROnPlateau](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html) provides better results and this is the following call i made:
 
` self.scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(self.optimizer, mode='max', factor=0.75,
                                                                        min_lr=1e-6,threshold=0.005,threshold_mode='abs',patience=10) `

Additionally, starting with a learning rate of 0.001 seems more effective. If you’d like, I can open a pull request to incorporate these changes.
